File: 3559-number-of-ways-to-assign-edge-weights-ii/3559-number-of-ways-to-assign-edge-weights-ii.java
class Solution {
    static int[] binaryExponent;
    static final long mod = 1000000007l;
    static {
        binaryExponent = new int[100001];
        long pro = 1l;
        for (int i = 1; i < binaryExponent.length; i++) {
            binaryExponent[i] = (int) (pro);
            pro = (pro * 2l) % mod;
        }
    }
    int[][] ancestors;
    int[] ans, parent;
    int[] depth;
    boolean[] visited;
    private void buildAncestors(int[] parent, int numNodes) {
        ancestors = new int[(int) (Math.log(numNodes) / Math.log(2)) + 1][numNodes + 1];
        int n = ancestors.length, m = ancestors[0].length;
        for (int i = 0; i < m; i++) {
            ancestors[0][i] = parent[i];
        }
        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                ancestors[i][j] = ancestors[i - 1][ancestors[i - 1][j]];
            }
        }
    }
    public int[] assignEdgeWeights(int[][] edges, int[][] queries) {
        int n = edges.length, q = queries.length;
        ans = new int[q];
        parent = new int[n + 2];
        visited = new boolean[n + 2];
        depth = new int[n + 2];
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i <= n + 1; i++) {
            adj.add(new ArrayList<>());
        }
        for (int i = 0; i < edges.length; i++) {
            adj.get(edges[i][0]).add(edges[i][1]);
            adj.get(edges[i][1]).add(edges[i][0]);
        }
        parent[1] = 1;
        dfs(1, adj, 0);
        buildAncestors(parent, n + 1);   
        for (int i = 0; i < q; i++) {
            ans[i] = binaryExponent[getPathCount(queries[i][0], queries[i][1])];
        }
        return ans;
    }
    private int getPathCount(int u, int v) {
        int depthU = depth[u], depthV = depth[v], lca = -1;
        if (depthU > depthV) lca = LCA(u, v);
        else lca = LCA(v, u);
        return depth[u] + depth[v] - 2 * depth[lca];
    }
    private int LCA(int u, int v) {
        int depthU = depth[u];
        int depthV = depth[v];
        int diff = depthU - depthV;
        int mask = 1, idx = 0;
        while (mask <= diff) {
            if ((mask & diff) > 0) {
                u = ancestors[idx][u];
            }
            mask <<= 1;
            idx++;
        }
        if (u == v) return u;
        int maxJumps = ancestors.length - 1;
        while (maxJumps >= 0) {
            if (ancestors[maxJumps][u] != ancestors[maxJumps][v]) {
                u = ancestors[maxJumps][u];
                v = ancestors[maxJumps][v];
            }
            maxJumps--;
        }
        int lca = ancestors[0][u];
        return lca;
    }
    private void dfs(int node, List<List<Integer>> adj, int depthValue) {
        visited[node] = true;
        depth[node] = depthValue;
        List<Integer> children = adj.get(node);
        for (int v : children) {
            if (visited[v]) continue;
            parent[v] = node;
            dfs(v, adj, depthValue + 1);
        }
    }
}

File: 3559-number-of-ways-to-assign-edge-weights-ii/3559-number-of-ways-to-assign-edge-weights-ii.cpp
class Solution {
public:
    static const int MOD = 1000000007;
    static vector<int> binaryExponent;
    vector<vector<int>> ancestors;
    vector<int> parent;
    vector<int> depth;
    vector<int> ans;
    void dfs(int node, int par, vector<vector<int>>& adj, int depthValue) {
        parent[node] = par;
        depth[node] = depthValue;
        for (int child : adj[node]) {
            if (child == par) continue;
            dfs(child, node, adj, depthValue + 1);
        }
    }
    void buildAncestors(int n) {
        int maxLog = 18; 
        ancestors.assign(maxLog, vector<int>(n + 1));
        for (int node = 1; node <= n; node++) {
            ancestors[0][node] = parent[node];
        }
        for (int jump = 1; jump < maxLog; jump++) {
            for (int node = 1; node <= n; node++) {
                ancestors[jump][node] = ancestors[jump - 1][ancestors[jump - 1][node]];
            }
        }
    }
    int LCA(int u, int v) {
        if (depth[u] < depth[v]) {
            swap(u, v);
        }
        int diff = depth[u] - depth[v];
        for (int bit = 0; diff > 0; bit++) {
            if (diff & 1) {
                u = ancestors[bit][u];
            }
            diff >>= 1;
        }
        if (u == v) return u;
        for (int jump = (int)ancestors.size() - 1;
             jump >= 0;
             jump--) {
            if (ancestors[jump][u] !=
                ancestors[jump][v]) {
                u = ancestors[jump][u];
                v = ancestors[jump][v];
            }
        }
        return ancestors[0][u];
    }
    int getPathLength(int u, int v) {
        int lca = LCA(u, v);
        return depth[u] + depth[v] - 2 * depth[lca];
    }
        vector<int> assignEdgeWeights(
        vector<vector<int>>& edges,
        vector<vector<int>>& queries) {
        int n = edges.size() + 1;
        int q = queries.size();
        parent.assign(n + 1, 0);
        depth.assign(n + 1, 0);
        ans.resize(q);
        vector<vector<int>> adj(n + 1);
        for (auto& edge : edges) {
            int u = edge[0];
            int v = edge[1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        dfs(1, 1, adj, 0);
        buildAncestors(n);
        for (int i = 0; i < q; i++) {
            int u = queries[i][0];
            int v = queries[i][1];
            int pathLength = getPathLength(u, v);
            ans[i] = binaryExponent[pathLength];
        }
        return ans;
    }
};
vector<int> Solution::binaryExponent = []() {
    vector<int> power(100001);
    long long curr = 1;

    for (int i = 1; i <= 100000; i++) {
        power[i] = curr;
        curr = (curr * 2LL) % Solution::MOD;
    }

    return power;
}();